### ISSUE TITLE
Any CPU, net461: ikvm-native.dll is not present

### ISSUE THREAD

**Issue by joniles** [ikvm-revived/ikvm#249]
When "Any CPU" is selected as the build target and using `net461`, no copy of `ikvm-native.dll` is written to the output directory, so code which uses JNI fails with the exception shown below. Selecting a specific CPU architecture (doesn't matter which one) resolves the issue. This is not an issue for `netcoreapp3.1`.

With "classic" IKVM this didn't appear to be an issue, so I'm assuming that one of the versions of `ikvm-native.dll` was selected as  the default?

(The example below is from the test case attached to https://github.com/ikvm-revived/ikvm/issues/248)

```
C:\Users\Jon\source\repos\ikvm-sqlite-jni\ikvm-sqlite-jni\bin\Debug\net461> .\ikvm-sqlite-jni.exe

Unhandled Exception: java.lang.ExceptionInInitializerError: System.DllNotFoundException: Unable to load DLL 'ikvm-native': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
   at IKVM.Runtime.JNI.Native.ikvm_GetJNIEnvVTable()
   at IKVM.Runtime.JNI.VtableBuilder..cctor()
   at java.lang.ClassLoader.loadLibrary0(Class fromClass, File file)
   at java.lang.ClassLoader.loadLibrary(Class fromClass, String name, Boolean isAbsolute)
   at java.lang.Runtime.load0(Class fromClass, String filename)
   at java.lang.System.load(String filename, CallerID )
   at org.sqlite.SQLiteJDBCLoader.loadNativeLibrary(String path, String name) in SQLiteJDBCLoader.java:line 300
   at org.sqlite.SQLiteJDBCLoader.extractAndLoadLibraryFile(String libFolderForCurrentOS, String libraryFileName, String  targetFolder) in SQLiteJDBCLoader.java:line 257
   at org.sqlite.SQLiteJDBCLoader.loadSQLiteNativeLibrary() in SQLiteJDBCLoader.java:line 366
   at org.sqlite.SQLiteJDBCLoader.initialize() in SQLiteJDBCLoader.java:line 69
   at org.sqlite.core.NativeDB.load() in NativeDB.java:line 63
   at org.sqlite.SQLiteConnection.open(String url, String origFileName, Properties props) in SQLiteConnection.java:line 241
   at org.sqlite.SQLiteConnection..ctor(String url, String fileName, Properties prop) in SQLiteConnection.java:line 61
   at org.sqlite.jdbc3.JDBC3Connection..ctor(String url, String fileName, Properties prop) in JDBC3Connection.java:line 28
   at org.sqlite.jdbc4.JDBC4Connection..ctor(String url, String fileName, Properties prop) in JDBC4Connection.java:line 22
   at org.sqlite.JDBC.createConnection(String url, Properties prop) in JDBC.java:line 115
   at net.sf.mpxj.common.SQLite.createConnection(File file, Properties props) in SQLite.java:line 78
   at net.sf.mpxj.common.SQLite.createConnection(File file) in SQLite.java:line 58
   at net.sf.mpxj.reader.UniversalProjectReader.populateSqliteTableNames(File file) in UniversalProjectReader.java:line 874
   at net.sf.mpxj.reader.UniversalProjectReader.handleSQLiteFile(InputStream stream) in UniversalProjectReader.java:line 548
   at net.sf.mpxj.reader.UniversalProjectReader.readInternal(File file, InputStream inputStream) in UniversalProjectReader.java:line 365
   at net.sf.mpxj.reader.UniversalProjectReader.handleFile(File file) in UniversalProjectReader.java:line 610
   at net.sf.mpxj.reader.UniversalProjectReader.readInternal(File file) in UniversalProjectReader.java:line 160
   at net.sf.mpxj.reader.UniversalProjectReader.readAll(File file) in UniversalProjectReader.java:line 144
   at net.sf.mpxj.reader.UniversalProjectReader.readAll(String fileName) in UniversalProjectReader.java:line 131
   at ikvm_sqlite_jni.Program.Main(String[] args)
C:\Users\Jon\source\repos\ikvm-sqlite-jni\ikvm-sqlite-jni\bin\Debug\net461>
```

**Comment by wasabii** [comment]
This is using the nuget package?

**Comment by joniles** [comment]
> This is using the nuget package?

Yes, same setup as https://github.com/ikvm-revived/ikvm/issues/248, `MavenReference` and IKVM 8.4.2.

**Comment by wasabii** [comment]
Working on the #248.

When building your project this way, in bin/, is there a runtimes/win7-*/native directory with ikvm-native in it? There should be multiple directories.

**Comment by wasabii** [comment]
Alrighty. Tracked down.

So, the IKVM.targets file distributed along with the NuGet package got left off at some point. Excluded from .gitignore, because it's inside a directory named build/.

The problem is like this: NuGet handles the runtimes\ directory automatically. For Core TFMs, it handles it correctly: copying the entire runtimes\**\native\*.* structure to the output, regardless of the specific RuntimeIdentifier. The .deps.json file is then assembled properly in order to point to specific libraries for specific RIDs. This is because the runtime is capable of making a decision, at runtime, as to the current RID, and then using the .deps.json file to track down the right library. All of files need to be included, regardless of the platform you're building for, because you can always just copy the output to another platform and run it there.

This is different from the Framework logic, which has no such resolution logic. Except for Mono. Mono does have such logic, but it's completely different than in Core. Mono has .dllmaps. But NuGet isn't smart enough to handle that.

So, when building for a Framework TFM, Windows is used as an assumption by NuGet. At build time the specific target is chosen, and only the single file is copied to the output. Hence, for any of that to work, at build time, it has to know the target RID. The Target RID is assumed from Platform if it can be.

The Core logic here is correct: always copy everything, then annotate what is needed for what. Because the actual platform we're executing might be unknown at build time, but only known at runtime.

The Framework logic has a Windows assumption, which is incorrect, since they can be run on Linux/OSX, using Mono. And also 32/64 bit versions of Windows. And Unity. Etc.

So, we have to specially handle the Framework logic. What we do here is ignore the copy of ikvm-native.dll which is copied directly into bin, and instead distribute our own NuGet targets that take effect for a Framework TFM, and copy the entire runtimes/ directory, and then at runtime have our own custom logic that resolves out of that based on the current platform.

The targets was missing.

Building another release with it done properly.

This whole problem will go away in 8.5, as we've completely removed ikvm-native.

**Comment by wasabii** [comment]
Fixed in 8.4.4

**Comment by joniles** [comment]
Thank you! Tested and working as expected here.
